platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java

```diff
@@ -368,4 +394,90 @@ private int findJoystickDevice(int device_id) {
 
 		return -1;
 	}
+
+	/**
+	 *
+	 * @param event
+	 * @param isCaptured - must be used and not MOUSE_MODE_CAPTURED because requestPointerCapture is not atomic and can make a race condition on fast mouse movement right after request
+	 */
+	private boolean handleMouseEvent(MotionEvent event, final boolean isCaptured) {
+		if (event.getActionMasked() == MotionEvent.ACTION_HOVER_ENTER ||
+				event.getActionMasked() == MotionEvent.ACTION_HOVER_MOVE ||
+				event.getActionMasked() == MotionEvent.ACTION_HOVER_EXIT) {
+			final int x = Math.round(event.getX());
+			final int y = Math.round(event.getY());
+			final int type = event.getAction();
+			queueEvent(new Runnable() {
+				@Override
+				public void run() {
+					GodotLib.hover(type, x, y);
+				}
+			});
+			return true;
+		} else if (event.getAction() == MotionEvent.ACTION_BUTTON_PRESS || event.getAction() == MotionEvent.ACTION_BUTTON_RELEASE) {
+			final float x = event.getX();
+			final float y = event.getY();
+			final int buttonMask = event.getButtonState();
+			final int action = event.getAction();
+			queueEvent(new Runnable() {
+				@Override
+				public void run() {
+					GodotLib.mouseEvent(action, buttonMask, x, y, isCaptured);
+				}
+			});
+			return true;
```

Comment:
This is not needed. A mouse `ACTION_BUTTON_PRESS` and `ACTION_BUTTON_RELEASE` events are also followed by `ACTION_DOWN` and `ACTION_UP` event in the `onTouchEvent` method. As such, we can just use the `GodotLib.touch` method and add the source and button mask to simplify the processing in the native code.

Reply:
I understand, but also, it feels like process_touch (on the cpp side) is very tailored to touch and gesture, should I just call the process mouse event in the jni side, or somehow merge the two in the cpp?

Reply:
Yeah, we can just call `process_mouse` on the jni side.

Reply:
just found out that clicking another mouse button doesn't fire `ACTION_DOWN`... for some reason Android sends `ACTION_MOVE` if one button is already clicked so there is no way to know if 2 buttons are clicked together... why?

Reply:
it doesn't happen in `onGenericMotionEvent`...